Modern UI - Save and restore current settings menu state when auto-hiding

123mpozzi opened the pull request:
## Description
<!-- Add a short description about the changes -->

Improves the auto-hide behavior when the settings panel is open, addressing UX issues where the menu could vanish too quickly.

### Changes

- use a longer inactivity timeout if the settings panel is left open without interaction
- ensure that when controls reappear (on mouse move or interaction), the settings panel is restored in the same state it was left in
  - restore the last-used navigation menu
  - restore the scroll position

### UX Considerations

Three possible approaches for handling auto-hide with the settings panel were explored:
1. Keep auto-hide active with the same rules
 - ✅ consistent with the rest of the UI
 - ❌ frustrating if the menu disappears while browsing options
 - ❌ users may feel  "punished" for not moving quickly enough.
2. Disable auto-hide while settings panel is open
- ✅ removes frustration: panel stays until dismissed and user stays in control
- ❌ risk of clutter if the user opens the settings and moves away
3. Hybrid (chosen): longer auto-hide delay while interacting with settings and remember what was opened
- ✅ panel stays open while interacting
- ✅ longer inactivity timeout (5–10× normal) still prevents clutter
- ✅  opened menus can reappear on any interaction, with the same state
- ❌ slightly more complex, but best balance of usability + cleanliness

We went with hybrid because it balances both goals (no frustration with quick auto-hide, no cluttering with timer removed), delivering a smoother, more predictable user experience.


## Checklist (for PR submitter and reviewers)
<!-- A PR with CHANGELOG entry will be released automatically -->
<!-- This is required and should be added in every case -->
- [x] `CHANGELOG` entry

123mpozzi committed:
Save and restore which settings panel was open before auto-hiding

- and use a longer auto-hide delay when Settings panel is open

123mpozzi committed:
Save and restore scroll state on auto-hide for settings panel

123mpozzi committed:
Extract settings panel auto-hide behavior

123mpozzi committed:
Do not hardcode the extended auto-hide delay

123mpozzi committed:
Add new line at end of file

123mpozzi committed:
Add settingsPanel auto-hide UX updates

stonko1994 reviewed (commented):
I quite like the idea and the basic UX of this. However, a few feedback points:
- The Settings Panel resizes when moving the mouse out and in again
- The 10 seconds are too long in case you keep the mouse in the view. I'd go for something around 3-5 seconds.
- The 'stored' state should restore after X second when the UI isn't going to be visible again soon. Maybe after X seconds after the UI hides, if no interaction is happening again in that timeframe, we clear the state

123mpozzi committed:
Reduce auto-hide layout to `4` seconds

123mpozzi committed:
Use correct panel size when restoring after auto-hiding

123mpozzi committed:
Clear the saved state after 15s of inactivity

123mpozzi commented:
Thanks for the feedback! 🙌 
 
> * The Settings Panel resizes when moving the mouse out and in again

Nice find! In b95d4d0007b39bb30fb182b7bde94b3d75a9f782 I implemented store/restore of panel width and height as well

> * The 10 seconds are too long in case you keep the mouse in the view. I'd go for something around 3-5 seconds.

I reduced the auto-hide delay to 4 seconds in 68873ab6238d1ecae82cd6d1bb9266a9bfd01600

> * The 'stored' state should restore after X second when the UI isn't going to be visible again soon. Maybe after X seconds after the UI hides, if no interaction is happening again in that timeframe, we clear the state

I agree. In 4cd8119b541ba8cbff6ee9184b9bec1b341015ba I apply a `15` seconds timeout after the component hides - after that timeout, the state is cleared

123mpozzi committed:
Update entry with state clearing after longer inactivity

123mpozzi committed:
Merge branch 'develop' into feature/modern-ui-save-and-restore-current-settings-menu-state-with-auto-hide

123mpozzi committed:
Move entry to unreleased section

stonko1994 commented on src/ts/components/settings/SettingsPanel.ts:
```diff
@@ -459,4 +459,39 @@ export class SettingsPanel<Config extends SettingsPanelConfig> extends Container
+    // Restore scroll positions after DOM is ready
+    setTimeout(() => {
+      this.getDomElement().get(0).scrollTop = scrollTop;
+      const wrapper = this.getDomElement().find('.bmpui-container-wrapper').get(0);
```
`bmpui` should not be hard-coded as it can be customized.

123mpozzi replied:
Address in 6f1ff90337abe64133ea9beec8279a7dd12a156c

stonko1994 replied:
wouldn't it be easier to expose the `innerContainerElement` from `Container` instead of trying to 'find' it again? 🤔 It will be always there and it might be a bit cleaner on the usage side.

You could just write `this.innerContainerElement` here? (we just need to make the `innerContainerElement` `protected` instead of `private`). Then we can also revert all of those `getWrapperClassName` which feels wrong to me because it looks a bit like a hack (our component structure shouldn't need to do class matching to access stuff)

123mpozzi replied:
Yes, that change makes a lot of sense! Address in 9bf4f077bc20778aa9f332abdc7234f687d32494

stonko1994 commented on src/ts/components/settings/SettingsPanel.ts:
```diff
@@ -459,4 +459,39 @@ export class SettingsPanel<Config extends SettingsPanelConfig> extends Container
+    scrollTop: number,
+    wrapperScrollTop: number,
+    panelWidth?: number,
+    panelHeight?: number,
```
Nit: Those are quite some properties already. Should we combine them into a helper struct? 🤔

123mpozzi replied:
Address in 6f1ff90337abe64133ea9beec8279a7dd12a156c

stonko1994 commented on src/ts/components/settings/SettingsPanelAutoHideManager.ts:
```diff
@@ -0,0 +1,150 @@
+/**
+ * State interface for preserving settings panel navigation and scroll position
+ */
+interface SettingsPanelState {
```
Oh the interface already exists (follow up from comment above)

123mpozzi replied:
Ah right! Changed the code to use the interface in 6f1ff90337abe64133ea9beec8279a7dd12a156c

stonko1994 commented on src/ts/components/settings/SettingsPanelAutoHideManager.ts:
```diff
@@ -0,0 +1,150 @@
+
+  constructor(
+    private uimanager: UIInstanceManager,
+    stateClearDelay: number = 15000,
```
To me the 15 seconds feel quite high. I played around with the value and I think that around 5 seconds feel about right to me.

Nit: Ideally we put that on the `SettingsPanelConfig` as customization option.

123mpozzi replied:
Moved the property under the config in 52e4a662ee4f7c71beedc610ae9b92f9c126695e

123mpozzi replied:
Actually, `SettingsPanelConfig` is the wrong place.

`SettingsPanelAutoHideManager` manages multiple settings panel with one delay, and the delay is not about individual panels, but about how long the UI remembers state after hiding.
  
  I think the right place is `UIConfig` since `stateClearDelay` is a global UI-level behavior. It applies to every UI instance.
  
  --- 
  
  I moved it again in 17cdc2b390bf547de07b79d447998515019656aa

123mpozzi replied:
> To me the 15 seconds feel quite high. I played around with the value and I think that around 5 seconds feel about right to me.

  I've changed the delay to 5 seconds in 7b324b404185b6049d5d3fa3ce124f9d3b9e9b2f

stonko1994 commented on src/ts/components/settings/SettingsPanelAutoHideManager.ts:
```diff
@@ -0,0 +1,150 @@
+      panel.show();
+
+      // Then restore the navigation state (this will override the resetNavigation call in onShow)
+      setTimeout(() => {
```
Why do we need this async after 0?

123mpozzi replied:
Good catch! `setTimeout(.., 0)` is a common pattern when you need to defer execution until after event handlers have fully completed. Without it, we'd be fighting against the panel's own `onShow` logic.

 It ensures the order of operations when showing the panel:

  1. `panel.show()` triggers the panel's `onShow` event handler
  2. The `SettingsPanel`'s `onShow` handler calls `resetNavigation(true)`, which:
    - Clears the navigation stack
    - Resets to the root page
    - Updates CSS classes and DOM dimensions

  If we called `panel.restoreNavigationState(state)` synchronously right after `panel.show()`, it would execute in the same call stack as the `resetNavigation()` call, potentially causing conflicts with DOM updates and animations that are still in progress.

  The `setTimeout(.., 0)` pushes the restoration to the next event loop tick.

  
  Sources:
  - https://imdeveshshukla.medium.com/settimeout-with-time-0-explained-967f3133e305
  - https://stackoverflow.com/a/779785
  - https://developer.mozilla.org/en-US/docs/Web/API/Window/setTimeout#delay

stonko1994 replied:
Would it be possible to add another `.onShow` listener to the panel? This can be compared with an `DispatchQueue.async.main` in iOS which we try to only use when absolutely necessary. Code like this is very prone for race-conditions so we should try to avoid it at any costs.

123mpozzi replied:
That's a really valid concern. Address in 751a664577bd8ba249e0055cb62b7ba224658c28

stonko1994 commented on src/ts/components/settings/SettingsPanelAutoHideManager.ts:
```diff
@@ -0,0 +1,150 @@
+
+    const openPanel = this.openSettingsPanels[0];
+    return openPanel.getConfig().hideDelay;
+  }
```
This method feels a bit weird to me. It only checks if we have an open settings panel and us the `SettingsPanel's` hide delay. Though the name really doesn't indicate that.

What do you think about the following:
- rename this to something like `getOpenSettingsPanelHideDelay`
- return an optional number
- use the base delay on the caller side as fallback if returning nil here.

123mpozzi replied:
Address in 84c24cdfec9b13f2d75696693ef671ce1a56bcaf

stonko1994 commented on src/ts/components/UIContainer.ts:
```diff
@@ -126,18 +130,28 @@ export class UIContainer extends Container<UIContainerConfig> {
+        // Use extended timer if settings panel is open, normal timer otherwise
+        const hideDelay = this.settingsPanelManager.getExtendedDelay(config.hideDelay);
+        this.uiHideTimeout.clear();
+        this.uiHideTimeout = new Timeout(hideDelay, this.hideUi);
```
I think assigning a new `Timeout` here is super dangerous. After this code path was executed for the first time we basically change the hide timout of our complete UI 👀 It is never reset to the original hide delay without an open settings panel.

123mpozzi replied:
We would still need to assign a new `TimeOut` since its `delay` is read-only by design.
In 84c24cdfec9b13f2d75696693ef671ce1a56bcaf I implemented the reset to base hide delay when no settings panel is opened

stonko1994 replied:
Comment here: https://github.com/bitmovin/bitmovin-player-ui/commit/84c24cdfec9b13f2d75696693ef671ce1a56bcaf#r167619568

123mpozzi replied:
Thanks for linking the comment! Address in 0b342a5b6e41f2858855ac95405b3ccc0faf02b3

stonko1994 commented on src/ts/components/UIContainer.ts:
```diff
@@ -327,6 +341,8 @@ export class UIContainer extends Container<UIContainerConfig> {
       }
     });
+
+    // Settings panel auto-hide management is now handled by SettingsPanelAutoHideManager
```
This comment is unnecessary.

123mpozzi replied:
Removed in 52e4a662ee4f7c71beedc610ae9b92f9c126695e

stonko1994 commented on src/ts/components/UIContainer.ts:
```diff
@@ -62,6 +63,8 @@ export class UIContainer extends Container<UIContainerConfig> {
   private userInteractionEvents: { name: string; handler: EventListenerOrEventListenerObject }[];
   private hidingPrevented: () => boolean;
 
+  private settingsPanelManager: SettingsPanelAutoHideManager;
```
Any particular reason you put that on the `UIContainer` and not on the `UIManager`? 🤔

123mpozzi replied:
I've initially placed it here because only `UIContainer` has auto-hide logic that needs extended delays and state restoration.

However, I see it make sense to move in `UIManager` since we follow a similar pattern for subtitle manager - Address in 1ea0918ff59058b30d74a2d4d75f896c275b58b0

stonko1994 reviewed (commented):
Manual testing outcome:
- The UX feels really really nice when tweaking the timeout.
- Issues:
  - I found an issue when resizing the window (basically when switching UIVariant) while a SettingsPanel is open. On my device the UI starts to flicker and no settings panel will open anymore.

https://github.com/user-attachments/assets/3149ddf8-e7f5-4cf3-91f1-436cd0a09f43

123mpozzi committed:
Move state clear delay to `SettingsPanelConfig`

123mpozzi committed:
Reuse settings panel state interface

And do not hardcore 'bmpui' prefix

123mpozzi committed:
Remove hardcoded prefix

123mpozzi committed:
Reset to base ui hide delay if no settings panel is open

123mpozzi committed:
Move `stateClearDelay` into UI container

Since it is a UI-level behavior, like `hideDelay`. Both control auto-hide timing:
  - `hideDelay`: how long before UI hides
  - `stateClearDelay`: how long UI remembers state after hiding

123mpozzi committed:
Change default state clear delay to 6 seconds

And fix double default values

123mpozzi committed:
Extract fields into config

123mpozzi committed:
Change default state clear delay to 5 seconds

123mpozzi committed:
Add type safeguard

We know that, at runtime, `stateClearDelay` will have its
default value of `5000`. However, there is a type mismatch
for Typescript: `stateClearDelay` is still optional in the
config

123mpozzi committed:
Fix `SettingsPanelAutoHideManager` being inconsistent with UI switches

123mpozzi committed:
Rescope `settingsPanelManager` since it handles multiple UI instances

123mpozzi commented:
> Manual testing outcome:
> 
> * The UX feels really really nice when tweaking the timeout.
> * Issues:
>   
>   * I found an issue when resizing the window (basically when switching UIVariant) while a SettingsPanel is open. On my device the UI starts to flicker and no settings panel will open anymore.
> 
>  ui-settings-panel.mov

Thanks for the video! 🙌 

I see, issue is `SettingsPanelAutoHideManager` is tracking panels across UI variant switches and trying to restore panels from a different UI instance, and different UI variants have different default auto-hide delay for the settings panel in `UIFactory`

It should be fixed as of 3fbe45d061f1c00295db9d6c39fc72fb1481286a

123mpozzi committed:
Add missing stub

123mpozzi committed:
Add `stateClearDelay` entry

123mpozzi committed:
Use more concise wording

123mpozzi committed:
Fix default value in docs

123mpozzi committed:
Remove unnecessary field

123mpozzi committed:
Rename settings panel auto hide manager vars

123mpozzi committed:
Fix settings panel flickering when using smallscreen UI variant

Issue was:
- settings panel is flickering between shown/hidden
- this results in the settings panel not being visible at all
And it is caused due to smallScreen UI giving `-1` as a auto-hide
value in `UIFactory`

123mpozzi committed:
Rename settings panel state retention config field

123mpozzi committed:
Cover custom hide delay and state persistence with stubs

123mpozzi committed:
Use real instances instead of stubs

123mpozzi committed:
Remove unnecessary state

123mpozzi committed:
Expose `innerContainerElement` instead of finding the element again

123mpozzi committed:
Remove debugging statements

123mpozzi committed:
Prevent usage of `setTimeout(.., 0)`

Since it could be good ground for race-conditions

123mpozzi committed:
Merge branch 'develop' into feature/modern-ui-save-and-restore-current-settings-menu-state-with-auto-hide

stonko1994 reviewed (commented):
As discussed offline let's change the approach to not try to apply the state retro-active but don't reset it in first place.

123mpozzi commented:
> As discussed offline let's change the approach to not try to apply the state retro-active but don't reset it in first place.

I've refactored in a new PR, #785 , since the changes were mostly independent
